### PR TITLE
Upgrade Python packages, switch to FastAPI lifespan async context manager

### LINKED ISSUE #3764: Many Python packages are outdated and need updating
Some python packages in the solution are old and have security issues. As many have dependencies on each other suggest bringing them all to newer versions where possible.

### PR DESCRIPTION
**Pull request by marrobi**
# Resolves #3764 #3405

- Updated Python packages where newer versions exist and don;t cause dependency conflicts. 
- Fixed tests where httpsx output has changed
- Removed depreciated startup and shutdown events and switched to FastAPI's lifespan async context
- Cosmos `query_items` no longer support enable_cross_partition_query. Have removed, functionality seems okay. Would be good if someone can confirm that this is appropriate action.
- Upgraded az cli in dev container to match python packages
- Pinned pip version in devcontainer

### CODE REVIEW

**Review by tamirkamara — COMMENTED** [review]
You mentioned security issues in the packages, so the question would be - when didn't dependabot trigger on those?

File: api_app/requirements.txt [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
-opencensus-ext-azure==1.1.7
+azure-core==1.29.5
+aiohttp==3.8.6
+azure-cosmos==4.5.1
+azure-identity==1.14.1
+azure-mgmt-cosmosdb==9.3.0
+azure-mgmt-compute==30.3.0
+azure-mgmt-costmanagement==4.0.1
+azure-storage-blob==12.18.3
+azure-servicebus==7.11.3
```
[the remark this answers is not in the document]
Comment (marrobi) [reply api_app/requirements.txt:10]: Ping me offline if want to discuss, were a couple in defender for cloud.

I've confirmed the API is good, you are right the RP is giving some noise. I will add them to disabled loggers, had done it as part of this draft PR: https://github.com/microsoft/AzureTRE/pull/3762/files#diff-e5c30790fe735c7619face05da76e2e5ab0afc9261155cc4a7239983e425e693

[the remark this answers is not in the document]
Comment (marrobi) [reply api_app/requirements.txt:10]: I've done that, the only things I've got now is:

```
_session_async azure.servicebus._pyamqp.aio._session_async 84813   2023-10-25 15:45:43,168 ERROR   Unable to attach new link: ValueError('Invalid link')
```

Which I have a support case open for.

**Review by martinpeck — APPROVED** [review]
I've looked at the FastAPI changes (context manager) and the change related to the enable_cross_partition_query change. Both of these look good to me. 

Approval given for those areas only. 
Not had a chance to look at anything else.

File: api_app/requirements.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (marrobi) [reply api_app/requirements.txt:10]: It is odd that see this error in the Resource Processor, but not the API.

[the remark this answers is not in the document]
Comment (marrobi) [reply api_app/requirements.txt:10]: Ah, the logging is configured differently.

However, also get the error when running this sample: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/servicebus/azure-servicebus/samples/async_samples/receive_queue_async.py

[the remark this answers is not in the document]
Comment (marrobi) [reply api_app/requirements.txt:10]: Same issue here - https://github.com/Azure/azure-sdk-for-python/issues/30907#issuecomment-1627691668 - any ideas what this is saying as to the issue and how to resolve?

[the remark this answers is not in the document]
Comment (marrobi) [reply api_app/requirements.txt:10]: Looks like this PR might help us when merged... https://github.com/Azure/azure-sdk-for-python/pull/30832

[the remark this answers is not in the document]
Comment (marrobi) [reply api_app/requirements.txt:10]: I've removed the looked for now, and will keep an eye on the PR:
```
    # Remove these once the following PR is merged:
    # https://github.com/Azure/azure-sdk-for-python/pull/30832
    # Issue: https://github.com/microsoft/AzureTRE/issues/3766
    "azure.servicebus._pyamqp.aio._session_async"
]
```

**Pull request merged**